pkg/chunk/storage/caching_storage_client.go

```diff
@@ -104,102 +113,101 @@ func newCachingStorageClient(client chunk.StorageClient, cache cache.Cache, vali
 	}
 }
 
-func (s *cachingStorageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch) (shouldContinue bool)) error {
-	value, ok, err := s.cache.Fetch(ctx, queryKey(query))
-	if err != nil {
-		cacheCorruptErrs.Inc()
-	}
+func (s *cachingStorageClient) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error {
+	// We cache the entire row, so filter client side.
+	callback = chunk_util.QueryFilter(callback)
+	cacheableMissed := []chunk.IndexQuery{}
+	missed := map[string]chunk.IndexQuery{}
+
+	span, ctx := ot.StartSpanFromContext(ctx, "Index cache lookups")
+	for _, query := range queries {
+		key := queryKey(query)
```

Comment:
Hmm, so `queryKey` only depends on `TableName` and `HashKey`. So if two different queries have those two same, we're in trouble. Though it might not be the case now (which I highly doubt), future schemas might make this point moot. Hence, I think we should make they key here out of everything rather than just `TableName` and `HashKey`.